File: Sources/ZarliAdapterAdMob/ZarliAdMobMediationAdapter.m
#import "ZarliAdMobMediationAdapter.h"
#import "ZarliAdMobMediationInterstitialAd.h"
#import "ZarliAdMobMediationRewardedAd.h"
@import ZarliSDKSwift;

@implementation ZarliAdMobMediationAdapter {
  ZarliAdMobMediationInterstitialAd *_interstitialAd;
  ZarliAdMobMediationRewardedAd *_rewardedAd;
}

#pragma mark - GADMediationAdapter Protocol

+ (void)setUpWithConfiguration:(GADMediationServerConfiguration *)configuration
             completionHandler:
                 (GADMediationAdapterSetUpCompletionBlock)completionHandler {
  // Zarli SDK initialization is handled globally via ZarliSDK.shared
  // No additional setup required for the adapter
  completionHandler(nil);
}

+ (GADVersionNumber)adSDKVersion {
  // Return Zarli SDK version
  GADVersionNumber version = {1, 3, 63};
  return version;
}

+ (GADVersionNumber)adapterVersion {
  // Return adapter version
  GADVersionNumber version = {1, 3, 63};
  return version;
}

+ (nullable Class<GADAdNetworkExtras>)networkExtrasClass {
  return nil;
}

#pragma mark - Interstitial Ad Loading

- (void)loadInterstitialForAdConfiguration:
            (GADMediationInterstitialAdConfiguration *)adConfiguration
                         completionHandler:
                             (GADMediationInterstitialLoadCompletionHandler)
                                 completionHandler {
  _interstitialAd = [[ZarliAdMobMediationInterstitialAd alloc]
      initWithAdConfiguration:adConfiguration
            completionHandler:completionHandler];
  [_interstitialAd loadAd];
}

#pragma mark - Rewarded Ad Loading

- (instancetype)init {
  self = [super init];
  if (self) {
    NSLog(@"[ZarliAdapter] ZarliAdMobMediationAdapter initialized");
  }
  return self;
}

- (void)loadRewardedAdForAdConfiguration:
            (GADMediationRewardedAdConfiguration *)adConfiguration
                       completionHandler:
                           (GADMediationRewardedLoadCompletionHandler)
                               completionHandler {
  _rewardedAd = [[ZarliAdMobMediationRewardedAd alloc]
      initWithAdConfiguration:adConfiguration
            completionHandler:completionHandler];
  [_rewardedAd loadAd];
}

@end

File: Sources/ZarliAdapterAdMob/ZarliAdMobMediationAdapter.swift
import Foundation
import GoogleMobileAds
import ZarliSDKSwift

@objc(ZarliAdMobMediationAdapter)
public class ZarliAdMobMediationAdapter: NSObject, GADMediationAdapter {
    
    // Properties to retain the ads during the loading process
    private var interstitialAd: ZarliAdMobMediationInterstitialAd?
    private var rewardedAd: ZarliAdMobMediationRewardedAd?
    
    public static func adapterVersion() -> GADVersionNumber {
        let version = GADVersionNumber(majorVersion: 1, minorVersion: 0, patchVersion: 0)
        return version
    }
    
    public static func adSDKVersion() -> GADVersionNumber {
        let version = GADVersionNumber(majorVersion: 1, minorVersion: 0, patchVersion: 0)
        return version
    }
    
    public static func networkExtrasClass() -> GADAdNetworkExtras.Type? {
        return nil
    }
    
    required public override init() {
        super.init()
    }
    
    public func setUp(with configuration: GADMediationServerConfiguration, completionHandler: @escaping GADMediationAdapterSetUpCompletionBlock) {
        
        // In this simple implementation, we assume readiness or successful setup.
        completionHandler(nil)
    }
    
    public func loadInterstitial(for adConfiguration: GADMediationInterstitialAdConfiguration, completionHandler: @escaping GADMediationInterstitialLoadCompletionHandler) {
        let ad = ZarliAdMobMediationInterstitialAd(configuration: adConfiguration, completionHandler: completionHandler)
        self.interstitialAd = ad // Retain
        ad.loadAd()
    }
    
    public func loadRewardedAd(for adConfiguration: GADMediationRewardedAdConfiguration, completionHandler: @escaping GADMediationRewardedLoadCompletionHandler) {
        let ad = ZarliAdMobMediationRewardedAd(configuration: adConfiguration, completionHandler: completionHandler)
        self.rewardedAd = ad // Retain
        ad.loadAd()
    }
}
